src/int_x64_backend.ml

```diff
@@ -7,24 +7,36 @@ let neg x = (-x)
 let add a b = a + b
 let sub a b = a - b
 let mul a b = a * b
-let div a b = a / b
+
+let unsigned_compare n m =
+  let open Nativeint in
+  compare (sub n min_int) (sub m min_int)
+
+let unsigned_div n d =
+  let open Nativeint in
+  if d < zero then
+    if unsigned_compare n d < 0 then zero else one
+  else
+    let q = shift_left (div (shift_right_logical n 1) d) 1 in
+    let r = sub n (mul q d) in
+    if unsigned_compare r d >= 0 then succ q else q
+
+let div a b = Nativeint.(to_int (unsigned_div (of_int a) (of_int b)))
 let rem a b = a mod b
 let succ x = x + 1
 let pred x = x - 1
-let abs x = Pervasives.abs x
+let abs x = if x >= 0 then x else (-x)
 let max_int = Int32.(to_int max_int)
 let min_int = Int32.(to_int min_int)
 let logand a b = a land b
 let logor a b = a lor b
 let logxor a b = a lxor b
-let lognot x = lnot (x land 0xffffffff)
+let lognot x = lnot x
 let shift_left a n = a lsl n
 let shift_right a n = a lsr n
 let shift_right_logical a n = a asr n
```

Comment:
`shift_right_logical` should be `lsr` (logical shift right) and `shift_right` should be `asr` (arithmetic shift right); they're switched here.